cluster-up/cluster/kind-k8s-sriov-1.17.0/config_sriov.sh

```diff
@@ -25,85 +22,134 @@ SRIOVDP_RESOURCE_NAME="sriov_net"
 VFS_DRIVER="vfio-pci"
 VFS_DRIVER_KMODULE="vfio_pci"
 
-function move_sriov_pfs_netns_to_node {
-  local -r node=$1
-  local -r pf_count_per_node=$2
-  local -r pid="$(docker inspect -f '{{.State.Pid}}' $node)"
-  local pf_array=()
+function prepare_node_netns() {
+  local -r node_name=$1
+  local -r node_pid=$(docker inspect -f '{{.State.Pid}}' "$node_name")
+
+  # Docker does not create the required symlink for a container netns
+  # it perverts iplink from learning that container netns.
+  # Thus it is necessary to create symlink between the current
+  # worker node (container) netns to /var/run/netns (consumed by iplink)
+  # Now the node container netns named with the node name will be visible.
+  ln -sf "/proc/$node_pid/ns/net" "/var/run/netns/$node_name"
+}
+
+function move_pf_to_node_netns() {
+  local -r node_name=$1
+  local -r pf_name=$2
 
-  mkdir -p /var/run/netns/
-  ln -sf /proc/$pid/ns/net "/var/run/netns/$node"
+  # Move PF to node network-namespace
+  ip link set "$pf_name" netns "$node_name"
+  # Ensure current PF is up
+  ip netns exec "$node_name" ip link set up dev "$pf_name"
+  ip netns exec "$node_name" ip link show
+}
 
+function get_pfs_names() {
   local -r sriov_pfs=( $(find /sys/class/net/*/device/sriov_numvfs) )
-  [ "${#sriov_pfs[@]}" -eq 0 ] && echo "FATAL: Could not find available sriov PFs" >&2 && return 1
+  [ "${#sriov_pfs[@]}" -eq 0 ] && echo "FATAL: Could not find available sriov PFs on host" >&2 && return 1
 
   for pf in "${sriov_pfs[@]}"; do
-    local pf_name="${pf%%/device/*}"
+    pf_name="${pf%%/device/*}"
     pf_name="${pf_name##*/}"
-
     if [ $(echo "${PF_BLACKLIST[@]}" | grep "${pf_name}") ]; then
       continue
     fi
 
-    # In case two clusters started at the same time, they might race on the same PF.
-    # The first will manage to assign the PF to its container, and the 2nd will just skip it
-    # and try the rest of the PFs available.
-    if ip link set "$pf_name" netns "$node"; then
-      if timeout 10s bash -c "until ip netns exec $node ip link show $pf_name > /dev/null; do sleep 1; done"; then
-        pf_array+=("$pf_name")
-        [ "${#pf_array[@]}" -eq "$pf_count_per_node" ] && break
-      fi
-    fi
+    pfs_names+=( $pf_name )
   done
 
-  [ "${#pf_array[@]}" -lt "$pf_count_per_node" ] && \
-    echo "FATAL: Not enough PFs allocated, PF_BLACKLIST (${PF_BLACKLIST[@]}), PF_COUNT_PER_NODE ${PF_COUNT_PER_NODE}" >&2 && \
-    return 1
+  echo "${pfs_names[@]}"
+}
 
-  echo "${pf_array[@]}"
+function configure_nodes_sriov_pfs_and_vfs() {
+  local -r nodes_array=($1)
+  local -r pfs_names_array=($2)
+  local -r pf_count_per_node=$3
+
+  mkdir -p /var/run/netns
```

Comment:
This is surprising to have it here. Is this intentional?

Reply:
Yes, you prefer this to be outside this function?

Reply:
I do not know, I think I prefer it close to the place it is actually needed. Here I do not understand where it is needed.

Reply:
We need to make sure /var/run/netns exists so that `lplink` will will "learn" the network namespaces of each worker node(container).
There is an explanation inside `prepare_node_netns` that mentioned this directory, but yeah no explanation for this line. 

I will add a comment next to it.

Reply:
So if you need it for `prepare_node_netns`, place it before the for-loop. Now it is between variables declarations.

Reply:
Done